### DISCUSSION TITLE
Mapping polymorphic effects

### DISCUSSION THREAD

**Discussion by chtenb** [koka-lang/koka#600] [Q&A]
```koka
pub effect yield<a>
  ctl yield(elem : a) : ()
  
pub fun map(it : () -> <yield<a>|e> (), f : a -> b) : <yield<b>|e> ()
  with override ctl yield(elem)
    yield(f(elem))
    resume(())
  it()
```

This does not work

```
repro.kk(6, 3): type error: abstract types do not match
  context        :   with override ctl yield(elem)
                       yield(f(elem))
                       resume(())
  term           :   with override ctl yield(elem)
                       yield(f(elem))
                       resume(())
  inferred effect: <yield<$a>,yield<$a>|_e1>
  expected effect: <yield<$a>,yield<$b>|_e>
  hint           : an higher-rank type escapes its scope?
```

If we replace the type `b` with `a`, it does work. 
I'm unsure why the inferred effect is `<yield<$a>,yield<$a>|_e1>`, because clearly the result of `f` is `b`.
Why does this fail, and how can we make it work?

**Comment by TimWhiting** [comment]
Interesting, I assumed that you could only override the "same" handler type. 

Looking at this issue I then wondered if something strange was going on with override that made that assumption - that the effect it masks is the exact same as the one it is defining. 
But even the desugaring doesn't work. 

```
pub effect yield<a>
  ctl yield(elem : a) : ()
  
pub fun map(it : () -> <yield<a>|e> (), f : a -> b) : <yield<b>|e> ()
  with handler
    ctl yield(elem)
      yield(f(elem))
      resume(())
  mask behind<yield>{it()}
```

In other words even if override should mean that the exact same effect at the same types is overridden, the desugaring should work to mask the effect regardless of type, and it doesn't seem to be doing so.
